src/PartKeepr/PartBundle/Entity/Part.php

```diff
@@ -823,24 +823,40 @@ public function recomputeStockLevels()
     {
         $sum = 0;
         $price = 0;
+		
+		$totalPartStockPrice = 0;
+        $lastPosEntryQuant = 0;
+        $lastPosEntryPrice = 0;
 
         foreach ($this->getStockLevels() as $stockLevel) {
-
-            if ($stockLevel->getStockLevel() < 0) {
-                $this->setRemovals(true);
-            } else {
-                $price += $stockLevel->getPrice() * $stockLevel->getStockLevel();
-            }
-            $sum += $stockLevel->getStockLevel();
-        }
-
-        if ($sum > 0) {
-            $this->setAveragePrice($price / $sum);
-        } else {
-            $this->setAveragePrice(0);
-        }
+			
+			$sum += $stockLevel->getStockLevel();
+			
+			if ($stockLevel->getStockLevel() > 0) {
+				$lastPosEntryQuant = $stockLevel->getStockLevel();
+				$lastPosEntryPrice = $stockLevel->getPrice();
+				$totalPartStockPrice += $lastPosEntryPrice * $lastPosEntryQuant;
+				$price = $totalPartStockPrice / $sum;
+			}
+			else {
+				if  ($sum < 0) {
+					$price = 0;
+					//$sum = 0;
+				}
+				else
+					if ($sum < $lastPosEntryQuant){
+						$totalPartStockPrice = $sum * $lastPosEntryPrice;
+						$price = $totalPartStockPrice / $sum;
+					}
+					else {
+						$totalPartStockPrice += $stockLevel->getStockLevel() * $price;
+						$price = $totalPartStockPrice / $sum;
+					}
+			}
+		}
 
         $this->setStockLevel($sum);
+		$this->setAveragePrice($price);
```

Comment:
Indentation wrong, please follow [PSR-2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) for coding standards. You can use the [php-cs-fixer](http://cs.sensiolabs.org/) to fix the coding style.